fennel/test_lib/data_engine.py

```diff
@@ -0,0 +1,648 @@
+import copy
+import logging
+from collections import defaultdict
+from dataclasses import dataclass
+from functools import partial
+from typing import Dict, List, Optional, Tuple, Callable, Any
+
+import numpy as np
+import pandas as pd
+from frozendict import frozendict
+
+import fennel.datasets.datasets
+from fennel.datasets import Dataset, Pipeline
+from fennel.datasets.datasets import sync_validation_for_pipelines
+from fennel.gen.dataset_pb2 import CoreDataset
+from fennel.lib.schema import data_schema_check
+from fennel.lib.to_proto import dataset_to_proto
+from fennel.sources import sources, PreProcValue
+from fennel.test_lib.executor import Executor
+from fennel.test_lib.test_utils import (
+    FakeResponse,
+    cast_df_to_schema,
+    cast_col_to_dtype,
+)
+import fennel.gen.schema_pb2 as schema_proto
+
+TEST_PORT = 50051
+TEST_DATA_PORT = 50052
+FENNEL_LOOKUP = "__fennel_lookup_exists__"
+FENNEL_ORDER = "__fennel_order__"
+FENNEL_TIMESTAMP = "__fennel_timestamp__"
+
+logger = logging.getLogger(__name__)
+
+
+def _transform_df(
```

Comment:
this is only for preproc right? lets call it _preproc_df ?

Reply:
chnaged.